In [47]:
import pandas as pd
import pathlib
import statsmodels.api as sm
from sklearn.preprocessing import LabelEncoder

In [48]:
path = pathlib.Path().cwd().parents[1] / 'CSVs' / 'Kings County Housing.csv'
df = pd.read_csv(path)
df.head()

id   price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront   
0   1  221900         3       1.00         1180      5650     1.0           0  \
1   2  538000         3       2.25         2570      7242     2.0           0   
2   3  180000         2       1.00          770     10000     1.0           0   
3   4  604000         4       3.00         1960      5000     1.0           0   
4   5  510000         3       2.00         1680      8080     1.0           0   

   view  sqft_basement  yr_built  yr_renovated  zipcode  
0     0              0      1955             0    98178  
1     0            400      1951          1991    98125  
2     0              0      1933             0    98028  
3     0            910      1965             0    98136  
4     0              0      1987             0    98074

- a) Where is King County? Use the zip codes if you are unsure.
    - It is in Washington

- b) How many observations are in the dataset? What does 1 row correspond to?
    - There are 21613 observations in the dataset
    - Each row corresponds to a house in the county

In [49]:
df.shape

(21613, 13)

- c) What are the median statistics for price, bedrooms, bathrooms, square foot of living space, and year built?
    - Price: 450000
    - Bedrooms: 3
    - Bathrooms: 2.25
    - Square Foot of Living Space: 1910
    - Year Built: 1975


In [50]:
df[['price', 'bedrooms', 'bathrooms', 'sqft_living', 'yr_built']].describe()

price      bedrooms     bathrooms   sqft_living      yr_built
count  2.161300e+04  21613.000000  21613.000000  21613.000000  21613.000000
mean   5.400881e+05      3.370842      2.114757   2079.899736   1971.005136
std    3.671272e+05      0.930062      0.770163    918.440897     29.373411
min    7.500000e+04      0.000000      0.000000    290.000000   1900.000000
25%    3.219500e+05      3.000000      1.750000   1427.000000   1951.000000
50%    4.500000e+05      3.000000      2.250000   1910.000000   1975.000000
75%    6.450000e+05      4.000000      2.500000   2550.000000   1997.000000
max    7.700000e+06     33.000000      8.000000  13540.000000   2015.000000

- d) Run the regression: $$P rice = a + b ∗Bedrooms$$
    - Write a full sentence explaining the coefficient on bedrooms.
        - The coefficient on bedrooms is about $121716$ which means that every additional bedroom corresponds to an increase of house price by about $121716$ dollars.
    - Is the coefficient statistically significant? What is the 95% confidence interval on the coefficient on bedrooms? Interpret the interval.
    - If a house has 2 bedrooms, what does the one variable model predict the price will be?
        - Approximately $\$373234.61$
    - Is the relationship between bedrooms and price necessarily causal?
    - Interpert the $R^2$ value of this model.

In [51]:
X = df['bedrooms']
Y = df['price']

X_full = sm.add_constant(X)

model = sm.OLS(Y, X_full)
results = model.fit()

display(results.summary())
f'The equation of the line is: y = {results.params[0]:.2f} + {results.params[1]:.2f}x'

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.095
Model:                            OLS   Adj. R-squared:                  0.095
Method:                 Least Squares   F-statistic:                     2271.
Date:                Thu, 09 Nov 2023   Prob (F-statistic):               0.00
Time:                        13:08:53   Log-Likelihood:            -3.0652e+05
No. Observations:               21613   AIC:                         6.131e+05
Df Residuals:                   21611   BIC:                         6.131e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       1.298e+05   8931.866     14.533      0.000    1.12e+05    1.47e+05
bedrooms    1.217e+05   2554.304     47.651      0.000    1.17e+05    1.27e+05
==============================================================================
Omnibus:                    18859.406   Durbin-Watson:                   1.961
Prob(Omnibus):                  0.000   Jarque-Bera (JB):          1199044.960
Skew:                           3.904   Prob(JB):                         0.00
Kurtosis:                      38.644   Cond. No.                         14.2
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

'The equation of the line is: y = 129802.36 + 121716.13x'

In [52]:
results.params[0] + results.params[1] * 2

373234.6093419398

- e) Run the regression of price on bedrooms and living square footage: $$\text{Price} = a + b * \text{Bedrooms} + c * \text{Sqft\_living}$$
    - Write a full sentence explaining the coefficient on bedrooms. How has it changed? Why might it have changed?
    - How has the $R^2$ changed from the first model?
    - What does the model predict for the price of a 2 bedroom, 1000 square foot apartment?
    - What does the model predict for the price of a 3 bedroom, 1000 square foot apartment?


In [53]:
X = df[['bedrooms', 'sqft_living']]
Y = df['price']

X_full = sm.add_constant(X)

model = sm.OLS(Y, X_full)
results = model.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.507
Model:                            OLS   Adj. R-squared:                  0.507
Method:                 Least Squares   F-statistic:                 1.110e+04
Date:                Thu, 09 Nov 2023   Prob (F-statistic):               0.00
Time:                        13:08:53   Log-Likelihood:            -2.9997e+05
No. Observations:               21613   AIC:                         5.999e+05
Df Residuals:                   21610   BIC:                         6.000e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const        7.947e+04   6604.764     12.032      0.000    6.65e+04    9.24e+04
bedrooms    -5.707e+04   2308.223    -24.723      0.000   -6.16e+04   -5.25e+04
sqft_living   313.9487      2.337    134.314      0.000     309.367     318.530
==============================================================================
Omnibus:                    14423.033   Durbin-Watson:                   1.986
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           492253.321
Skew:                           2.732   Prob(JB):                         0.00
Kurtosis:                      25.732   Cond. No.                     8.87e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 8.87e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [54]:
display(results.params[0] + results.params[1] * 2 + results.params[2] * 1000)
results.params[0] + results.params[1] * 3 + results.params[2] * 1000

279284.5271091432

222217.76818621776

- f) Add dummies for zip code to your second model and run the regression: $$\text{Price} = a + b * \text{Bedrooms} + c * \text{Sqft\_living} + d * \text{Zip}$$ You should have 70 zip dummies. You do not need to interpret them, just include them.
    - What is the R2of this model? Write a full sentence.
    - What is the coefficient on bedrooms? How does it compare to the other models? Is it statistically significant?
    - Suppose we wanted to use this model to make a casual statement about the effect of bedrooms. Write a full sentence about the assumption we would have to make.


In [55]:
X = pd.concat([df[['bedrooms', 'sqft_living', 'yr_built']], pd.DataFrame({'zipcode': LabelEncoder().fit_transform(list(df['zipcode'].astype(str).values))})], axis=1)
Y = df['price']

X_full = sm.add_constant(X)

model = sm.OLS(Y, X_full)
results = model.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.541
Model:                            OLS   Adj. R-squared:                  0.541
Method:                 Least Squares   F-statistic:                     6370.
Date:                Thu, 09 Nov 2023   Prob (F-statistic):               0.00
Time:                        13:08:54   Log-Likelihood:            -2.9919e+05
No. Observations:               21613   AIC:                         5.984e+05
Df Residuals:                   21608   BIC:                         5.984e+05
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const        4.555e+06   1.25e+05     36.299      0.000    4.31e+06     4.8e+06
bedrooms    -5.953e+04   2232.906    -26.661      0.000   -6.39e+04   -5.52e+04
sqft_living   341.1127      2.354    144.916      0.000     336.499     345.726
yr_built    -2305.1922     63.499    -36.303      0.000   -2429.655   -2180.729
zipcode       565.9472     93.677      6.041      0.000     382.334     749.561
==============================================================================
Omnibus:                    13967.707   Durbin-Watson:                   1.978
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           472593.053
Skew:                           2.605   Prob(JB):                         0.00
Kurtosis:                      25.308   Cond. No.                     2.19e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.19e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

- g) Run one more model to evaluate the effect of bedrooms on price, picking some other variable(s) for controls. What variables did you include? Write the full estimating equation, and include a screenshot of your results. What coefficient for bedrooms do you find?

In [56]:
X = pd.concat([
    df[['bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors', 'sqft_basement', 'yr_built']], 
    pd.DataFrame({'zipcode': LabelEncoder().fit_transform(list(df['zipcode'].astype(str).values))})
    ], axis=1)

Y = df['price']

X_full = sm.add_constant(X)

model = sm.OLS(Y, X_full)
results = model.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.556
Model:                            OLS   Adj. R-squared:                  0.556
Method:                 Least Squares   F-statistic:                     3388.
Date:                Thu, 09 Nov 2023   Prob (F-statistic):               0.00
Time:                        13:08:54   Log-Likelihood:            -2.9882e+05
No. Observations:               21613   AIC:                         5.977e+05
Df Residuals:                   21604   BIC:                         5.977e+05
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
const          6.537e+06   1.45e+05     45.073      0.000    6.25e+06    6.82e+06
bedrooms      -6.806e+04   2248.323    -30.271      0.000   -7.25e+04   -6.37e+04
bathrooms      6.947e+04   3892.951     17.844      0.000    6.18e+04    7.71e+04
sqft_living     307.0594      3.211     95.641      0.000     300.766     313.352
sqft_lot         -0.3036      0.041     -7.328      0.000      -0.385      -0.222
floors         4.721e+04   4272.324     11.050      0.000    3.88e+04    5.56e+04
sqft_basement   -15.1318      4.968     -3.046      0.002     -24.870      -5.393
yr_built      -3359.4434     74.530    -45.075      0.000   -3505.528   -3213.359
zipcode         190.1724     95.168      1.998      0.046       3.636     376.709
==============================================================================
Omnibus:                    14350.430   Durbin-Watson:                   1.985
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           524048.664
Skew:                           2.687   Prob(JB):                         0.00
Kurtosis:                      26.517   Cond. No.                     3.85e+06
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.85e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""